### PR TITLE
Add null terminator for char* based json string unescape, and precondition on input length

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-c/issues/2301

### CODE REVIEW

File: sdk/tests/core/test_az_json.c [hunk 1]
```diff
@@ -3105,17 +3105,6 @@ static void test_az_json_string_unescape(void** state)
     assert_true(az_span_is_content_equal(expected, destination_span));
   }
 
-  // insufficient space
-  {
-    az_span json = AZ_SPAN_FROM_STR(" { \"name\": \"some value string\" , \"code\" : 123456 } ");
-    char destination[5] = { 0 };
-    int final_size;
-
-    az_result result = az_json_string_unescape(json, destination, 4, &final_size);
```
Comment (LarryOsterman) [inline sdk/tests/core/test_az_json.c:3114]: Strangely, this test for unescape doesn't have an escaped string.

I would have expected to see a test which had a string with escaped values where the destination was too short for the unescaped string.

File: sdk/src/azure/core/az_json_token.c [hunk 2]
```diff
@@ -333,7 +333,8 @@ AZ_NODISCARD az_result az_json_string_unescape(
 {
   _az_PRECONDITION_VALID_SPAN(json_string, 1, false);
   _az_PRECONDITION_NOT_NULL(destination);
-  _az_PRECONDITION(destination_max_size > 0);
```
Comment (LarryOsterman) [inline sdk/src/azure/core/az_json_token.c:336]: Interestingly, the precondition at line 335 precludes the more traditional use of this pattern:

```c
err = az_json_string_unescape(json_string, NULL, 0, &size);
char *dest = malloc(size);
err = az_json_string_unescape(json_string, dest, size, &size);
```

Instead, you need to do something silly like:
```c
char dummy;
err = az_json_string_unescape(json_string, &dummy, 0, &size);
char *dest = malloc(size);
err = az_json_string_unescape(json_string, dest, size, &size);
```

**Review by LarryOsterman — APPROVED** [review]

File: sdk/tests/core/test_az_json.c [hunk 3]
```diff
@@ -3105,17 +3105,6 @@ static void test_az_json_string_unescape(void** state)
     assert_true(az_span_is_content_equal(expected, destination_span));
   }
 
-  // insufficient space
-  {
-    az_span json = AZ_SPAN_FROM_STR(" { \"name\": \"some value string\" , \"code\" : 123456 } ");
```
Comment (antkmsft) [inline sdk/tests/core/test_az_json.c:3110]: We need to be testing for `AZ_ERROR_NOT_ENOUGH_SPACE`, `AZ_ERROR_UNEXPECTED_CHAR`, `AZ_ERROR_UNEXPECTED_END`, for the null terminator at the end, for the precondition, for the correct size being written to `out_string_length`, for the precondition, and for the fact that the string is indeed unescaped where expected.

Maybe some of it is being tested, but it is not obvious so far. So far I can see you are deleting this coverage (one instance is legitimately a duplicate?) But I do not see at least the test for null terminator and for the precondition.

**Review by antkmsft — APPROVED** [review]
Will look good once covered by tests.
Functionality LGTM, I see no issues.

File: sdk/tests/core/test_az_json.c (hunk 3 above)
Comment (ahsonkhan) [reply sdk/tests/core/test_az_json.c:3110]: The other cases you mentioned are covered already.

We haven't tested the other preconditions for this method, so we can improve coverage separately. The goal of the PR is just to fix the null terminator.

I will add the test case for nulls at the end.

File: sdk/tests/core/test_az_json.c (hunk 1 above)
Comment (ahsonkhan) [reply sdk/tests/core/test_az_json.c:3114]: Agreed. That said, this existing test was only testing destination size being too small, independent of unescaping.

File: sdk/src/azure/core/az_json_token.c (hunk 2 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_token.c:336]: Our existing SDK APIs don't behave like this, where the required size is returned on a failed call.

File: sdk/tests/core/test_az_json.c (hunk 3 above)
Comment (ahsonkhan) [reply sdk/tests/core/test_az_json.c:3110]: Here's the coverage. The null terminator line is covered:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=1769797&view=codecoverage-tab

And the `AZ_ERROR_NOT_ENOUGH_SPACE` case will only be possible if preconditions are off, which is why it isn't being covered now. We don't expect anyone to hit that case (and it is only there to avoid memory corruption, as an extra check).

I'll file an issue to address the test coverage for this API around preconditions, since the issue already exists in main.

Comment (ahsonkhan) [reply sdk/tests/core/test_az_json.c:3110]: https://github.com/Azure/azure-sdk-for-c/issues/2310

Comment (antkmsft) [reply sdk/tests/core/test_az_json.c:3110]: Is it even needed then, if it is a precondition? Once you violate the precondition, you're in the undefined behavior territory, and also the purposes of preconditions is for our code to not spend extra cycles checking for errors. I'd say it should be one or another.

Null terminator string will be shown as covered, but that's only because the line was executed during tests, and the tool can't see that you are not actually checking the buffer that null terminator was indeed being written. Please add that test.

File: sdk/src/azure/core/az_json_token.c [hunk 4]
```diff
@@ -333,7 +333,8 @@ AZ_NODISCARD az_result az_json_string_unescape(
...
+  // The destination needs to be larger than the input, for null terminator.
+  _az_PRECONDITION(destination_max_size > az_span_size(json_string));
 
   int32_t position = 0;
   int32_t span_size = az_span_size(json_string);
```
Comment (antkmsft) [inline sdk/src/azure/core/az_json_token.c:340]: Since size is precondition, `if (position > destination_max_size)` can be dropped.
It checks on every loop iteration.

Either way, even if we keep it, it would be much cheaper to just test it before the loop, just as you do with precondition.

Then, it could be a single `if (current_char == '\\') { if (i < span_size - 1) { ... } else { ... } }`, instead of checking `if (current_char == '\\')` twice (even if not for optimization, but readability as well).

Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_token.c:340]: You are right.

I think I am going to re-look at this whole implementation, including the optimizations you mentioned, as part of:
https://github.com/Azure/azure-sdk-for-c/issues/2302

I want to keep the code change in this PR minimal, for now, to avoid too much churn, since it will likely be throw away work.

File: sdk/src/azure/core/az_json_token.c (hunk 2 above)
Comment (LarryOsterman) [reply sdk/src/azure/core/az_json_token.c:336]: That's a shame, because the entire reason for this pattern `([in]foo*, [in]int fooSize, [out]int* sizeNeeded)` is to enable that use scenario (passing in 0 and `NULL` to do a sizing pass followed by the actual allocation).
